File: libs/adsp-service-sdk/src/registration/registration.ts
import axios from 'axios';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const retry = require('promise-retry');
import { Logger } from 'winston';
import { ServiceRegistrar, ServiceRegistration, ServiceRole } from './index';
import { TokenProvider } from '../access';
import { ServiceDirectory } from '../directory';
import { AdspId, adspId } from '../utils';

export class ServiceRegistrarImpl implements ServiceRegistrar {
  private readonly LOG_CONTEXT = { context: 'ServiceRegistration' };

  constructor(
    private readonly logger: Logger,
    private readonly directory: ServiceDirectory,
    private readonly tokenProvider: TokenProvider
  ) {}

  async register(registration: ServiceRegistration): Promise<void> {
    if (registration.configurationSchema) {
      const namespace = registration.serviceId.namespace;
      const name = registration.serviceId.service;
      const update = {
        [`${namespace}:${name}`]: {
          configurationSchema: registration.configurationSchema,
        },
      };
      await this.updateConfiguration(adspId`urn:ads:platform:configuration-service`, update);
    }

    if (registration.roles) {
      const update = {
        [registration.serviceId.toString()]: {
          roles: registration.roles.map((r) => ((r as ServiceRole).role ? r : { role: r, description: '' })) || [],
        },
      };
      await this.updateConfiguration(adspId`urn:ads:platform:tenant-service`, update);
    }

    if (registration.events) {
      const namespace = registration.serviceId.service;
      const update = {
        [namespace]: {
          name: namespace,
          definitions: registration.events.reduce(
            (defs, def) => ({
              ...defs,
              [def.name]: {
                name: def.name,
                description: def.description,
                payloadSchema: def.payloadSchema,
                interval: def.interval,
                log: def.log
              },
            }),
            {}
          ),
        },
      };
      await this.updateConfiguration(adspId`urn:ads:platform:event-service`, update);
    }

    if (registration.eventStreams) {
      const update = registration.eventStreams.reduce(
        (streams, stream) => ({
          ...streams,
          [stream.id]: {
            id: stream.id,
            name: stream.name,
            description: stream.description,
            events: stream.events,
            subscriberRoles: stream.subscriberRoles,
            publicSubscribe: stream.publicSubscribe,
          },
        }),
        {}
      );
      await this.updateConfiguration(adspId`urn:ads:platform:push-service`, update);
    }

    if (registration.fileTypes) {
      const update = registration.fileTypes.reduce(
        (types, type) => ({
          ...types,
          [type.id]: {
            id: type.id,
            name: type.name,
            anonymousRead: type.anonymousRead,
            readRoles: type.readRoles,
            updateRoles: type.updateRoles,
          },
        }),
        {}
      );
      await this.updateConfiguration(adspId`urn:ads:platform:file-service`, update);
    }

    if (registration.notifications) {
      const update = registration.notifications.reduce(
        (types, type) => ({
          ...types,
          [type.name]: {
            ...type,
            id: type.name,
            name: type.displayName || type.name,
            events: type.events,
          },
        }),
        {}
      );

      await this.updateConfiguration(adspId`urn:ads:platform:notification-service`, update);
    }

    if (registration.values) {
      const namespace = registration.serviceId.service;
      const update = {
        [namespace]: {
          name: namespace,
          definitions: registration.values.reduce(
            (defs, def) => ({
              ...defs,
              [def.id]: {
                name: def.id,
                displayName: def.name,
                description: def.description,
                jsonSchema: def.jsonSchema,
              },
            }),
            {}
          ),
        },
      };
      await this.updateConfiguration(adspId`urn:ads:platform:value-service`, update);
    }

    if (registration.serviceConfigurations) {
      for (const { serviceId, configuration } of registration.serviceConfigurations) {
        await this.updateConfiguration(serviceId, configuration);
      }
    }
  }

  private async updateConfiguration<C extends Record<string, unknown>>(serviceId: AdspId, update: C): Promise<void> {
    const configurationServiceId = adspId`urn:ads:platform:configuration-service:v2`;
    const serviceUrl = await this.directory.getServiceUrl(configurationServiceId);

    try {
      await retry(async (next, count) => {
        try {
          await this.#tryUpdateConfiguration(serviceUrl, count, serviceId, update);
        } catch (err) {
          this.logger.debug(`Try ${count} failed with error. ${err}`, this.LOG_CONTEXT);
          next(err);
        }
      });

      this.logger.info(`Updated registration configuration of ${serviceId}.`, this.LOG_CONTEXT);
    } catch (err) {
      this.logger.error(`Error encountered registering configuration. ${err}`);
      throw err;
    }
  }

  #tryUpdateConfiguration = async (
    configurationServiceUrl: URL,
    count: number,
    serviceId: AdspId,
    update: Record<string, unknown>
  ): Promise<void> => {
    this.logger.debug(`Try ${count}: updating configuration for service ${serviceId}...`, this.LOG_CONTEXT);

    const configurationUrl = new URL(
      `v2/configuration/${serviceId.namespace}/${serviceId.service}`,
      configurationServiceUrl
    );

    const token = await this.tokenProvider.getAccessToken();
    await axios.patch(
      configurationUrl.href,
      {
        operation: 'UPDATE',
        update,
      },
      { headers: { Authorization: `Bearer ${token}` } }
    );
  };
}

File: libs/adsp-service-flask-sdk/adsp_service_flask_sdk/registration.py
import logging
from typing import Any, Dict, List, NamedTuple, Optional

from httpx import RequestError, post

from ._constants import (
    PLATFORM_CONFIGURATION_API,
    PLATFORM_CONFIGURATION_SERVICE,
    PLATFORM_EVENT_SERVICE,
    PLATFORM_TENANT_SERVICE,
)
from .adsp_id import AdspId
from .directory import ServiceDirectory
from .token_provider import TokenProvider


class ConfigurationDefinition(NamedTuple):
    description: str
    schema: Dict[str, Any]


class ServiceRole(NamedTuple):
    role: str
    description: str
    in_tenant_admin: bool = False


class DomainEventDefinition(NamedTuple):
    name: str
    description: str
    payload_schema: Dict[str, Any]


class AdspRegistration(NamedTuple):
    display_name: Optional[str] = None
    description: Optional[str] = None
    health_endpoint_path: Optional[str] = None
    docs_endpoint_path: Optional[str] = None
    configuration: Optional[ConfigurationDefinition] = None
    roles: Optional[List[ServiceRole]] = []
    events: Optional[List[DomainEventDefinition]] = []


class ServiceRegistrar:
    _logger = logging.getLogger("adsp.service-registrar")

    def __init__(
        self,
        service_id: AdspId,
        directory: ServiceDirectory,
        token_provider: TokenProvider,
    ) -> None:
        self.__service_id = service_id
        self.__directory = directory
        self.__token_provider = token_provider

    def register(self, registration: AdspRegistration) -> None:
        if registration.configuration:
            self._updated_service_configuration(
                PLATFORM_CONFIGURATION_SERVICE,
                {
                    f"{self.__service_id.namespace}:{self.__service_id.service}": {
                        "description": registration.configuration.description,
                        "configurationSchema": registration.configuration.schema,
                    }
                },
            )

        if registration.roles:
            self._updated_service_configuration(
                PLATFORM_TENANT_SERVICE,
                {
                    str(self.__service_id): {
                        "roles": [
                            {
                                "role": role.role,
                                "description": role.description,
                                "inTenantAdmin": role.in_tenant_admin,
                            }
                            for role in registration.roles
                        ]
                    }
                },
            )

        if registration.events:
            self._updated_service_configuration(
                PLATFORM_EVENT_SERVICE,
                {
                    self.__service_id.service: {
                        "service": self.__service_id.service,
                        "definitions": {
                            event.name: {
                                "name": event.name,
                                "description": event.description,
                                "payloadSchema": event.payload_schema,
                            }
                            for event in registration.events
                        },
                    }
                },
            )

    def _updated_service_configuration(
        self, service_id: AdspId, update: Dict[str, Any]
    ):
        configuration_service_url = self.__directory.get_service_url(
            PLATFORM_CONFIGURATION_API
        )
        token = self.__token_provider.get_access_token()

        try:
            post(
                f"{configuration_service_url}/configuration/{service_id.namespace}/{service_id.service}",
                json={"operation": "UPDATE", "update": update},
                headers={"Authorization": f"Bearer {token}"},
            )

            self._logger.info("Updated registration configuration for service %s", service_id)
        except RequestError as err:
            self._logger.error(
                "Error encountered updating registration configuration for service %s. %s",
                service_id,
                err,
            )
            raise
